package/src/components/Message/hooks/useProcessReactions.ts

```diff
@@ -0,0 +1,132 @@
+import { DefaultStreamChatGenerics } from '../../../types/types';
+import { ReactionListProps } from '../MessageSimple/ReactionList';
+
+export type ReactionSummary = {
+  count: number;
+  firstReactionAt: Date | null;
+  Icon: ComponentType | null;
+  isOwnReaction: boolean;
+  lastReactionAt: Date | null;
+  latestReactedUserNames: string[];
+  type: string;
+  unlistedReactedUserCount: number;
+};
+
+export type ReactionsComparator = (a: ReactionSummary, b: ReactionSummary) => number;
+
+type UseProcessReactionsParams<
+  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
+> = Pick<
+  ReactionListProps<StreamChatGenerics>,
+  'own_reactions' | 'reaction_groups' | 'latest_reactions'
+> &
+  Pick<MessagesContextValue<StreamChatGenerics>, 'supportedReactions'> & {
+    sortReactions?: ReactionsComparator;
+  };
+
+export const defaultReactionsSort: ReactionsComparator = (a, b) => {
+  if (a.firstReactionAt && b.firstReactionAt) {
+    return +a.firstReactionAt - +b.firstReactionAt;
+  }
+
+  return a.type.localeCompare(b.type, 'en');
+};
+
+export const useProcessReactions = <
+  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
+>(
+  props: UseProcessReactionsParams<StreamChatGenerics>,
+) => {
+  const {
+    latest_reactions: propLatestReactions,
+    own_reactions: propOwnReactions,
+    reaction_groups: propReactionGroups,
+    sortReactions: propSortReactions,
+    supportedReactions: propSupportedReactions,
+  } = props;
+
+  const { message } = useMessageContext();
+  const { supportedReactions: contextSupportedReactions } = useMessagesContext();
+  const supportedReactions = propSupportedReactions || contextSupportedReactions;
+  const latestReactions = propLatestReactions || message.latest_reactions;
+  const ownReactions = propOwnReactions || message?.own_reactions;
+  const reactionGroups = propReactionGroups || message?.reaction_groups;
+  const sortReactions = propSortReactions || defaultReactionsSort;
+
+  const isOwnReaction = useCallback(
+    (reactionType: string) =>
+      ownReactions?.some((reaction) => reaction.type === reactionType) ?? false,
+    [ownReactions],
+  );
+
+  const getEmojiByReactionType = useCallback(
+    (reactionType: string) =>
+      supportedReactions.find(({ type }) => type === reactionType)?.Icon ?? null,
+    [supportedReactions],
+  );
+
+  const isSupportedReaction = useCallback(
+    (reactionType: string) =>
+      supportedReactions.some((reactionOption) => reactionOption.type === reactionType),
+    [supportedReactions],
+  );
+
+  const getLatestReactedUserNames = useCallback(
+    (reactionType?: string) =>
+      latestReactions?.flatMap((reaction) => {
+        if (reactionType && reactionType === reaction.type) {
+          const username = reaction.user?.name || reaction.user?.id;
+          return username ? [username] : [];
+        }
+        return [];
+      }) ?? [],
+    [latestReactions],
+  );
```

Comment:
As we are not exposing these functions, wouldn't it be better if they are defined within the scope of the `useMemo` closures? I think that way we'll simplify the code and we won't have to deal with `useCallback` and too many dependency arrays.

Reply:
Or even better, move them away from this hook:

```ts
const isSupportedReaction =  (supportedReactions: Reaction[], reactionType: string) =>
      supportedReactions.some((reactionOption) => reactionOption.type === reactionType);

const useProcessReactions = () => {
   // use the helpers here
};
```

Reply:
Done